emukit/model_wrappers/gpy_model_wrappers.py

```diff
@@ -134,7 +134,14 @@ def generate_hyperparameters_samples(self, n_samples=20, n_burnin=100, subsample
 
         """
         self.model.optimize(max_iters=self.n_restarts)
-        self.model.param_array[:] = self.model.param_array * (1. + np.random.randn(self.model.param_array.size) * 0.01)
+        # Add jitter to all unfixed parameters. After optimizing the hyperparameters, the gradient of the
+        # posterior probability of the parameters wrt. the parameters will be close to 0.0, which is a poor
+        # initialization for HMC
+        unfixed_params = [param for param in self.model.flattened_parameters if not param.is_fixed]
+        for param in unfixed_params:
+            # Add jitter by multiplying with log-normal noise with mean 1 and standard deviation 0.01 
+            # This ensures the sign of the parameter remains the same
+            param *= np.random.lognormal(np.log(1. / np.sqrt(1.0001)), np.sqrt(np.log(1.0001)), size=param.size)
```

Comment:
Yes, it does work because `param` is not a float or an array - it's an instance of a mutable type `GPy.core.parameterization.param.Param`:
```
In: type(model.flattened_parameters[0])
Out: GPy.core.parameterization.param.Param
```
Assigning an array/float to it changes it in place. Running the following snippet makes it clear:
```
In [3]: model = GPy.models.GPRegression(np.ones([3, 2]), np.ones([3, 1]), GPy.kern.Matern52(2, ARD=True))

In [4]: model["Mat52.variance"].fix()
Out[4]: array([0])

In [5]: model.param_array
Out[5]: array([1., 1., 1., 1.])

In [6]: # Add jitter to all unfixed parameters. After optimizing the hyperparameters, the gradient of the
   ...: # posterior probability of the parameters wrt. the parameters will be close to 0.0, which is a poor
   ...: # initialization for HMC
   ...: unfixed_params = [param for param in model.flattened_parameters if not param.is_fixed]
   ...: for param in unfixed_params:
   ...:     # Add jitter by multiplying with log-normal noise with mean 1 and standard deviation 0.01
   ...:     # This ensures the sign of the parameter remains the same
   ...:     param *= np.random.lognormal(np.log(1. / np.sqrt(1.0001)), np.sqrt(np.log(1.0001)), size=param.size)
   ...:

In [7]: model.param_array
Out[7]: array([1.        , 1.00995493, 1.01032573, 1.00097379])
```

Reply:
I personally also find this solution in GPy unintuitive, but after searching for quite a while, this the cleanest API call I found to get only the unfixed parameters without referencing private variables. Another way would be to call `model.parameter_names_flat(include_fixed=False)` to get the names of the viable parameters, but these names need to be sanitised first before they can be used to get the parameters as they're all prepended with `GPRegression.` (e.g. `GPRegression.Mat52.variance`)

Reply:
confusing decision in GPy, i agree. but so be it, if that works